pkg/cachemanager/cachemanager_test.go

```diff
@@ -0,0 +1,373 @@
+	require.NoError(t, cacheManager.UpsertSource(ctx, sourceB, []schema.GroupVersionKind{podGVK, configMapGVK}))
+
+	// ... expect the aggregator to dedup
+	require.True(t, cacheManager.gvksToSync.IsPresent(configMapGVK))
+	require.True(t, cacheManager.gvksToSync.IsPresent(podGVK))
+	require.ElementsMatch(t, cacheManager.watchedSet.Items(), []schema.GroupVersionKind{podGVK, configMapGVK})
+
+	// adding a source without a previously added gvk ...
+	require.NoError(t, cacheManager.UpsertSource(ctx, sourceB, []schema.GroupVersionKind{configMapGVK}))
+	// ... should not remove any gvks that are still referenced by other sources
+	require.True(t, cacheManager.gvksToSync.IsPresent(configMapGVK))
+	require.True(t, cacheManager.gvksToSync.IsPresent(podGVK))
+
+	// adding a source that modifies the only reference to a gvk ...
+	require.NoError(t, cacheManager.UpsertSource(ctx, sourceB, []schema.GroupVersionKind{nsGVK}))
+
+	// ... will effectively remove the gvk from the aggregator
+	require.False(t, cacheManager.gvksToSync.IsPresent(configMapGVK))
+	require.True(t, cacheManager.gvksToSync.IsPresent(podGVK))
+	require.True(t, cacheManager.gvksToSync.IsPresent(nsGVK))
+}
+
+// TestCacheManager_RemoveSource tests that we can modify the gvk aggregator when removing a source.
+func TestCacheManager_RemoveSource(t *testing.T) {
+	cacheManager, ctx := unitCacheManagerForTest(t)
+	configMapGVK := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "ConfigMap"}
+	podGVK := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Pod"}
+	sourceA := aggregator.Key{Source: "a", ID: "source"}
+	sourceB := aggregator.Key{Source: "b", ID: "source"}
+
+	// seed the gvk aggregator
+	require.NoError(t, cacheManager.gvksToSync.Upsert(sourceA, []schema.GroupVersionKind{podGVK}))
+	require.NoError(t, cacheManager.gvksToSync.Upsert(sourceB, []schema.GroupVersionKind{podGVK, configMapGVK}))
+
+	// removing a source that is not the only one referencing a gvk ...
+	require.NoError(t, cacheManager.RemoveSource(ctx, sourceB))
+	// ... should not remove any gvks that are still referenced by other sources
+	require.True(t, cacheManager.gvksToSync.IsPresent(podGVK))
+	require.False(t, cacheManager.gvksToSync.IsPresent(configMapGVK))
+
+	require.NoError(t, cacheManager.RemoveSource(ctx, sourceA))
+	require.False(t, cacheManager.gvksToSync.IsPresent(podGVK))
+}
+
+func newSyncExcluderFor(nsToExclude string) *process.Excluder {
+	excluder := process.New()
+	excluder.Add([]configv1alpha1.MatchEntry{
+		{
+			ExcludedNamespaces: []wildcard.Wildcard{wildcard.Wildcard(nsToExclude)},
+			Processes:          []string{"sync"},
+		},
+		// exclude kube-system by default to prevent noise
+		{
+			ExcludedNamespaces: []wildcard.Wildcard{"kube-system"},
+			Processes:          []string{"sync"},
+		},
+	})
+
+	return excluder
+}
+
+func unstructuredFor(gvk schema.GroupVersionKind, name string) *unstructured.Unstructured {
```

Comment:
I've found that a better pattern for doing this kind of thing is using "functional options": https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

We already have a thing in `pkg` that's a variant of that; https://github.com/open-policy-agent/gatekeeper/blob/master/pkg/fakes/pod.go

Why don't we move this there and make it more reusable.

Reply:
the `unstructureFor` funcs are neat convenience funcs to get back unstructured items, not like `v1.Pod`, to use in tests. Unless someone also feels strongly about this, I'd prefer we punt on adding more code since this PR is already large for now. Happy to add/ revisit out of band?